models/pull.go

```diff
@@ -270,64 +270,98 @@ func (pr *PullRequest) Merge(doer *User, baseGitRepo *git.Repository) (err error
-		fmt.Sprintf("PullRequest.Merge (git merge --no-ff --no-commit): %s", tmpBasePath),
-		"git", "merge", "--no-ff", "--no-commit", "head_repo/"+pr.HeadBranch); err != nil {
-		return fmt.Errorf("git merge --no-ff --no-commit [%s]: %v - %s", tmpBasePath, err, stderr)
+	// Run the merge algorithm throughout the prepared index.
+	if stdout, stderr, err = process.GetManager().ExecDirEnv(-1, "",
+		fmt.Sprintf("PullRequest.Merge (git merge-index)"),
+		[]string{
+			"GIT_DIR=" + baseRepoPath,
+			"GIT_INDEX_FILE=" + indexTmpPath,
+			"GIT_WORK_TREE=" + workTreeTmpPath,
+		},
+		"git", "merge-index", "git-merge-one-file", "-a"); err != nil {
+		return fmt.Errorf("git merge-index git merge-one-file -a: %s (%v)", stderr, err)
 	}
 
-	sig := doer.NewGitSig()
-	if _, stderr, err = process.GetManager().ExecDir(-1, tmpBasePath,
-		fmt.Sprintf("PullRequest.Merge (git merge): %s", tmpBasePath),
-		"git", "commit", fmt.Sprintf("--author='%s <%s>'", sig.Name, sig.Email),
-		"-m", fmt.Sprintf("Merge branch '%s' of %s/%s into %s", pr.HeadBranch, pr.HeadUserName, pr.HeadRepo.Name, pr.BaseBranch)); err != nil {
-		return fmt.Errorf("git commit [%s]: %v - %s", tmpBasePath, err, stderr)
+	// Write the tree object back to the Git object store from the index.
+	if stdout, stderr, err = process.GetManager().ExecDirEnv(-1, "",
+		fmt.Sprintf("PullRequest.Merge (git write-tree): %s", baseRepoPath),
+		[]string{"GIT_DIR=" + baseRepoPath, "GIT_INDEX_FILE=" + indexTmpPath},
+		"git", "write-tree"); err != nil {
+		return fmt.Errorf("git write-tree: %s (%v)", stderr, err)
 	}
+	treeID := strings.TrimSpace(stdout)
+
+	log.Trace("BaseRepo:%s TreeId:%s", baseRepoPath, treeID)
 
-	// Push back to upstream.
-	if _, stderr, err = process.GetManager().ExecDir(-1, tmpBasePath,
-		fmt.Sprintf("PullRequest.Merge (git push): %s", tmpBasePath),
-		"git", "push", baseGitRepo.Path, pr.BaseBranch); err != nil {
-		return fmt.Errorf("git push: %s", stderr)
+	// Create the commit object based on the tree object we just built.
+	var headBranch *Branch
+	if headBranch, err = pr.HeadRepo.GetBranch(pr.HeadBranch); err != nil {
+		return fmt.Errorf("Getting head branch: %s: %s (%v)", pr.HeadBranch, stderr, err)
 	}
 
+	var headCommit *git.Commit
+	if headCommit, err = headBranch.GetCommit(); err != nil {
+		return fmt.Errorf("Getting head commit: %s: %s (%v)", pr.HeadBranch, stderr, err)
+	}
+	commitMessage := fmt.Sprintf("Merge branch '%s' of %s/%s into %s", pr.HeadBranch, pr.HeadUserName, pr.HeadRepo.Name, pr.BaseBranch)
+
+	if stdout, stderr, err = process.GetManager().ExecDirEnv(-1, "",
+		fmt.Sprintf("PullRequest.Merge (git commit-tree): %s", baseRepoPath),
+		[]string{
+			"GIT_DIR=" + baseRepoPath,
+			"GIT_AUTHOR_NAME=" + headCommit.Author.Name,
+			"GIT_AUTHOR_EMAIL=" + headCommit.Author.Email,
+			"GIT_AUTHOR_DATE=" + headCommit.Author.When.Format("Mon, 02 Jan 2006 15:04:05 -0700"),
+			"GIT_COMMITTER_NAME=" + doer.NewGitSig().Name,
+			"GIT_COMMITTER_EMAIL=" + doer.NewGitSig().Email,
```

Comment:
Does this hide the Email if the commiter has that set in their config?

Reply:
Yes. https://sourcegraph.com/github.com/go-gitea/gitea@6c6533fb31bbc82e7907b1dc233212045849569b/-/blob/models/user.go#L170